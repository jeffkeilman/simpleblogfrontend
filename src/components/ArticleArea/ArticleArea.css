.ArticleArea sup {
  font-style: italic;
}

.ArticleArea h1 {
  margin-top: 0;
  font-size: 2em;
}

.Article {
  border-bottom: 1px solid #808080;
  padding: 10px 15px;
}