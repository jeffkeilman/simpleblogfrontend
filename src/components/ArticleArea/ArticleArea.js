import './ArticleArea.css'
import React from 'react'

const articles = [
  {
    id: 1,
    posted_dt: '2021-11-09T22:15:41.118Z',
    title: 'My first article',
    preview: "This is my first article. I'm writing it to learn about blogging. I would like to become a more proficient writer. I don't anticipate this blah blah yadda yadda doesn't matter just need some more length"
  },
  {
    id: 2,
    posted_dt: '2021-10-21T22:15:41.118Z',
    title: 'My second article',
    preview: "This is my second article. I'm writing it to learn about blogging. I would like to become a more proficient writer. I don't anticipate this blah blah yadda yadda doesn't matter just need some more length"
  },
  {
    id: 3,
    posted_dt: '2021-10-03T22:15:41.118Z',
    title: 'My third article',
    preview: "This is my third article. I'm writing it to learn about blogging. I would like to become a more proficient writer. I don't anticipate this blah blah yadda yadda doesn't matter just need some more length."
  }
]

class Description extends React.Component {
  trimDate (dateStr) {
    const dateArr = (
      dateStr // Tue Nov 09 2021
        .split(' ') // ['Tue', 'Nov', '09', '2021']
        .slice(1) // ['Nov', '09', '2021']
    )

    dateArr[1] = dateArr[1] + ',' // ['Nov', '09,', '2021']
    return dateArr.join(' ')
  }

  mapArticles (articles) {
    if (articles && articles.length) {
      // if there are any articles
      return articles.map(article => {
        const articleDateStr = new Date(article.posted_dt).toDateString()
        const articleDateStrTrimmed = this.trimDate(articleDateStr)
        const lastCharInPreview = article.preview[article.preview.length - 1]
        if (lastCharInPreview === '.') {
          // if last char in preview is a period, strip it from the preview so ... doesn't look weird
          article.preview = article.preview.substring(0, article.preview.length - 1)
        }
        return (
          <article key={article.id} className='Article'>
            <sup>Posted on: {articleDateStrTrimmed}</sup>
            <h1>{article.title}</h1>
            <p>{article.preview}...</p>
          </article>
        )
      })
    }
  }

  render () {
    const articleList = this.mapArticles(articles)
    return (
      <section className='ArticleArea'>
        {articleList}
      </section>
    )
  }
}

export default Description
