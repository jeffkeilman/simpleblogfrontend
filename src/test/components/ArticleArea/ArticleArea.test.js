import React from 'react'
import renderer from 'react-test-renderer'
import ArticleArea from '../../../components/ArticleArea/ArticleArea'
import { render } from '@testing-library/react'

describe('ArticleArea', () => {
  // snapshot test
  it('renders correctly', () => {
    const aa = renderer.create(<ArticleArea />).toJSON()
    expect(aa).toMatchSnapshot()
  })

  // TODO: mock actual API calls and test that real articles are created
  // for now, just test mocked articles
  it('renders articles', () => {
    const { container } = render(<ArticleArea />)
    const articles = container.querySelectorAll('.Article')
    // should have rendered 3 articles
    expect(articles.length).toBe(3)

    // check dates
    expect(articles[0].querySelector('sup').innerHTML).toBe('Posted on: Nov 09, 2021')
    expect(articles[1].querySelector('sup').innerHTML).toBe('Posted on: Oct 21, 2021')
    expect(articles[2].querySelector('sup').innerHTML).toBe('Posted on: Oct 03, 2021')

    // check titles
    expect(articles[0].querySelector('h1').innerHTML).toBe('My first article')
    expect(articles[1].querySelector('h1').innerHTML).toBe('My second article')
    expect(articles[2].querySelector('h1').innerHTML).toBe('My third article')

    // check article preview
    const article1Preview = articles[0].querySelector('p').innerHTML
    const article2Preview = articles[1].querySelector('p').innerHTML
    const article3Preview = articles[2].querySelector('p').innerHTML
    // these will be much smaller once mocked
    const expectedArticle1 = "This is my first article. I'm writing it to learn about blogging. I would like to become a more proficient writer. " +
      "I don't anticipate this blah blah yadda yadda doesn't matter just need some more length..."
    const expectedArticle2 = "This is my second article. I'm writing it to learn about blogging. I would like to become a more proficient writer. " +
      "I don't anticipate this blah blah yadda yadda doesn't matter just need some more length..."
    const expectedArticle3 = "This is my third article. I'm writing it to learn about blogging. I would like to become a more proficient writer. " +
      "I don't anticipate this blah blah yadda yadda doesn't matter just need some more length..."

    expect(article1Preview).toBe(expectedArticle1)
    expect(article2Preview).toBe(expectedArticle2)
    expect(article3Preview).toBe(expectedArticle3)
  })
})
