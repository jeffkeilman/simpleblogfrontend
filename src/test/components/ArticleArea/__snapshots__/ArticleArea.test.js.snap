// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleArea renders correctly 1`] = `
<section
  className="ArticleArea"
>
  <article
    className="Article"
  >
    <sup>
      Posted on: 
      Nov 09, 2021
    </sup>
    <h1>
      My first article
    </h1>
    <p>
      This is my first article. I'm writing it to learn about blogging. I would like to become a more proficient writer. I don't anticipate this blah blah yadda yadda doesn't matter just need some more length
      ...
    </p>
  </article>
  <article
    className="Article"
  >
    <sup>
      Posted on: 
      Oct 21, 2021
    </sup>
    <h1>
      My second article
    </h1>
    <p>
      This is my second article. I'm writing it to learn about blogging. I would like to become a more proficient writer. I don't anticipate this blah blah yadda yadda doesn't matter just need some more length
      ...
    </p>
  </article>
  <article
    className="Article"
  >
    <sup>
      Posted on: 
      Oct 03, 2021
    </sup>
    <h1>
      My third article
    </h1>
    <p>
      This is my third article. I'm writing it to learn about blogging. I would like to become a more proficient writer. I don't anticipate this blah blah yadda yadda doesn't matter just need some more length
      ...
    </p>
  </article>
</section>
`;
